pallets/pallet-bonded-coins/src/curves.rs

```diff
@@ -0,0 +1,260 @@
+	transcendental::{exp, ln, sqrt},
+};
+
+use crate::{PassiveSupply, Precision};
+
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub enum Curve<Parameter> {
+	Polynomial(PolynomialParameters<Parameter>),
+	SquareRoot(SquareRootParameters<Parameter>),
+	LMSR(LMSRFunctionParameters<Parameter>),
+}
+
+pub enum Operation<PassiveSupply> {
+	Mint(PassiveSupply),
+	Burn(PassiveSupply),
+}
+
+impl<Balance> Operation<PassiveSupply<Balance>> {
+	pub fn inner_value(&self) -> &PassiveSupply<Balance> {
+		match self {
+			Operation::Mint(x) => x,
+			Operation::Burn(x) => x,
+		}
+	}
+}
+
+impl<Parameter> Curve<Parameter>
+where
+	Parameter: FixedSigned + PartialOrd<Precision> + From<Precision> + ToFixed,
+	<Parameter as Fixed>::Bits: Copy + ToFixed + AddAssign + BitOrAssign + ShlAssign,
+{
+	fn calculate_accumulated_passive_issuance<Balance: Fixed>(passive_issuance: &[Balance]) -> Balance {
+		passive_issuance
+			.iter()
+			.fold(Balance::from_num(0), |sum, x| sum.saturating_add(*x))
+	}
+
+	fn calculate_integral_bounds(
+		op: Operation<PassiveSupply<Parameter>>,
+		active_issuance_pre: Parameter,
+		active_issuance_post: Parameter,
+	) -> (Parameter, Parameter) {
+		match op {
+			Operation::Burn(passive) => {
+				let accumulated_passive_issuance = Self::calculate_accumulated_passive_issuance(&passive);
+				(
+					active_issuance_post.saturating_add(accumulated_passive_issuance),
+					active_issuance_pre.saturating_add(accumulated_passive_issuance),
+				)
+			}
+			Operation::Mint(passive) => {
+				let accumulated_passive_issuance = Self::calculate_accumulated_passive_issuance(&passive);
+				(
+					active_issuance_pre.saturating_add(accumulated_passive_issuance),
+					active_issuance_post.saturating_add(accumulated_passive_issuance),
+				)
+			}
+		}
+	}
+	pub fn calculate_cost(
+		&self,
+		active_issuance_pre: Parameter,
+		active_issuance_post: Parameter,
+		op: Operation<PassiveSupply<Parameter>>,
+	) -> Result<Parameter, ArithmeticError> {
+		match self {
+			Curve::Polynomial(params) => {
+				let (low, high) = Self::calculate_integral_bounds(op, active_issuance_pre, active_issuance_post);
+
+				params.calculate_costs(low, high)
+			}
+			Curve::SquareRoot(params) => {
+				let (low, high) = Self::calculate_integral_bounds(op, active_issuance_pre, active_issuance_post);
+				params.calculate_costs(low, high)
+			}
+			Curve::LMSR(params) => {
+				let passive_issuance_over_e = op
+					.inner_value()
+					.iter()
+					.map(|x| params.calculate_passive_issuance(*x))
+					.collect::<Result<Vec<Parameter>, ArithmeticError>>()?;
+
+				let passive_issuance = passive_issuance_over_e
+					.iter()
+					.try_fold(Parameter::from_num(0), |acc, x| {
+						acc.checked_add(*x).ok_or(ArithmeticError::Overflow)
+					})?;
+
+				let lmsr_calc = LMSRCalculation {
+					m: params.m,
+					passive_issuance,
+				};
+				lmsr_calc.calculate_costs(active_issuance_pre, active_issuance_post)
```

Comment:
does this give us different results if we swap pre and post (ignoring the sign)? if not, we could just work with lower and higher as well.

Reply:
In lower and higher, the passive supply is added. This would corrupt the implementation.